src/snowflake/snowpark/relational_grouped_dataframe.py

```diff
@@ -197,11 +225,17 @@ def _to_df(self, agg_exprs: List[Expression]) -> DataFrame:
                 analyzer=self._df._session._analyzer,
             )
 
-        return DataFrame(self._df._session, group_plan)
+        return DataFrame(
+            self._df._session, group_plan, ast_stmt=_ast_stmt, _emit_ast=_emit_ast
+        )
 
     @relational_group_df_api_usage
+    @publicapi
     def agg(
-        self, *exprs: Union[Column, Tuple[ColumnOrName, str], Dict[str, str]]
+        self,
+        *exprs: Union[Column, Tuple[ColumnOrName, str], Dict[str, str]],
+        _ast_stmt: Optional[proto.Assign] = None,
```

Comment:
Some of these calls have _ast_stmt as a param and others only have emit_ast, is it that the ones with _ast_stmt can be called from other methods internally?

Also - rather than "worm-hole" these parameters through each an every call - can't you take advantage of the thread local to build the ast without changing the function signature on each and every interface? This is how a lot of tracing/span tools work - not sure if it applies here though.

Reply:
Neither _emit_ast nor _ast_stmt should be considered context – their value might reasonably change from call to call. We use `emit_ast` for both global suppression (where it would indeed be context-like) as well as for when we invoke public APIs from within other public APIs.

Reply:
`_emit_ast` is the more recent design. Originally, we started with `_ast_stmt`. Yet, having an explicit parameter `_emit_ast` is much better. However, sometimes in order to pass state properly (in terms of assigns), it is convenient to have `_ast_stmt` as well for accessor objects like RelationalGroupedDataFrame, DataframeNaFunctions, DataFrameWriter, .... It's a mix of history and making life easier.